### ISSUE TITLE
Discussion: 1 body per request vs multiple bodies per request

### ISSUE THREAD

**Issue by onbjerg** [paradigmxyz/reth#226]
Creating this issue to unblock #220 and continue the discussion here.

The question is: What are the tradeoffs between requesting 1 body per request to a peer vs requesting multiple bodies per request to a peer?

Currently:

- The body downloader requests 1 body per call to the P2P layer, but does multiple calls at the same time
- This is done because the downloader returns bodies in order of *block number*, not the order in which requests are fulfilled
- This means that the downloader assumes that the body returned is valid (but as soon as it is handed over to the stage, it is validated)

Alternatively:

- We request multiple bodies per call to the P2P layer
- There is no order guarantee on the bodies returned by the P2P layer, so we have to match ourselves
- A way to match is to calculate the transactions root of the body, however, in order to match this root with a header, we need to store a mapping of `TxRoot => HeaderHash`, and the downloader would need a RO transaction to the database (which is not the case currently)

In other words: Is the added complexity (extra database calls, more computation) worth the upside? What is the upside?

Worth noting:

- If we compute the transaction root as soon as the request is fulfilled, we might compute the transaction root for a large number of bodies we later discard, if they are after an invalid body. This is not currently the case, since the transaction root is calculated in the order the blocks are handed over to the stage
- The alternative requires a new table for only that part of sync (it is not used by e.g. RPC)

**Comment by onbjerg** [comment]
We are now doing multiple bodies per request
